File: problems-adapt/0001-0100/0079_prefix-tree/solution_trie.rs
struct TrieNode {
    children: [Option<Box<TrieNode>>; 26],
    end: bool,
}

impl TrieNode {
    fn new() -> Self {
        // One trie node: 26 child slots indexed by c - 'a' plus a
        // whole-word terminator flag; nodes appear lazily on insert.
        TrieNode {
            children: Default::default(),
            end: false,
        }
    }
}

pub struct PrefixTree {
    root: TrieNode,
}

impl PrefixTree {
    pub fn new() -> Self {
        PrefixTree { root: TrieNode::new() }
    }

    pub fn insert(&mut self, word: String) {
        let mut node = &mut self.root;
        for letter in word.bytes() {
            let slot = (letter - b'a') as usize;
            if node.children[slot].is_none() {
                node.children[slot] = Some(Box::new(TrieNode::new()));
            }
            node = node.children[slot].as_mut().unwrap();
        }
        node.end = true;
    }

    // Walks one node per character; None as soon as a slot is empty.
    fn walk(&self, text: &str) -> Option<&TrieNode> {
        let mut node = &self.root;
        for letter in text.bytes() {
            node = node.children[(letter - b'a') as usize].as_deref()?;
        }
        Some(node)
    }

    pub fn search(&mut self, word: String) -> bool {
        match self.walk(&word) {
            Some(node) => node.end,
            None => false,
        }
    }

    pub fn hasPrefix(&mut self, prefix: String) -> bool {
        self.walk(&prefix).is_some()
    }
}

File: problems-adapt/0201-0300/0208_prefix-tree/solution_trie.ts
interface TrieNode {
    children: (TrieNode | null)[];
    end: boolean;
}

class PrefixTree {
    private root: TrieNode = PrefixTree.makeNode();

    // One trie node: 26 child slots indexed by c - 'a' plus a whole-word
    // terminator flag; nodes appear lazily on insert.
    private static makeNode(): TrieNode {
        return { children: new Array<TrieNode | null>(26).fill(null), end: false };
    }

    insert(word: string): void {
        let node: TrieNode = this.root;
        for (const letter of word) {
            const slot = letter.charCodeAt(0) - 97;
            if (node.children[slot] === null) {
                node.children[slot] = PrefixTree.makeNode();
            }
            node = node.children[slot]!;
        }
        node.end = true;
    }

    // Walks one node per character; null as soon as a slot is empty.
    private walk(text: string): TrieNode | null {
        let node: TrieNode | null = this.root;
        for (const letter of text) {
            node = node.children[letter.charCodeAt(0) - 97];
            if (node === null) {
                return null;
            }
        }
        return node;
    }

    search(word: string): boolean {
        const node = this.walk(word);
        return node !== null && node.end;
    }

    hasPrefix(prefix: string): boolean {
        return this.walk(prefix) !== null;
    }
}
